src/main/java/com/microsoft/azure/vmagent/AzureVMAgent.java

```diff
@@ -526,6 +544,122 @@ public synchronized void deprovision(Localizable reason) throws Exception {
                 new Object[]{this.getDisplayName(), reason == null ? "Unknown reason" : reason.toString()});
 
         computer.setAcceptingTasks(false);
+
+        ComputerLauncher launcher = computer.getLauncher();
+
+        if ((launcher != null && launcher instanceof AzureVMAgentSSHLauncher)) {
+            AzureVMAgentSSHLauncher azureLauncher = (AzureVMAgentSSHLauncher) launcher;
+
+            OutputStream terminateResults = new OutputStream() {
+                private String terminateResultBuffer = "";
+
+                @Override
+                public void write(int b) throws IOException {
+                    terminateResultBuffer += (char) b;
+                }
+
+                @Override
+                public void flush() {
+                    if (!terminateResultBuffer.isEmpty()) {
+                        LOGGER.log(Level.FINER, "AzureVMAgent: deprovision: terminateScript, " + terminateResultBuffer);
+                        terminateResultBuffer = "";
+                    }
+                }
+            };
+
+            PrintStream terminateStream = new PrintStream(terminateResults, true, "UTF-8");
+
+            final boolean isUnix = this.getOsType().equals(OperatingSystemTypes.LINUX);
+            // Check if VM is already stopped or stopping or getting deleted ,
+            // if yes then there is no point in trying to connect
+            // Added this check - since after restarting jenkins master,
+            // jenkins is trying to connect to all the agents although agents are suspended.
+            // This still means that a delete agent will eventually get cleaned up.
+            try {
+                if (!this.isVMAliveOrHealthy()) {
+                    LOGGER.log(Level.INFO,
+                            "AzureVMAgent: deprovision: Agent {0} is shut down, deleted, etc. "
+                                    + "Not attempting to connect",
+                            computer.getName());
+                }
+            } catch (Exception e1) {
+                // ignoring exception purposefully
+            }
+
+            Thread loggingThread = new Thread(new Runnable() {
+                private static final int FLUSH_INTERVAL = 30000;
+                public void run() {
+                    try {
+                        while (terminateResults != null) {
```

Comment:
This condition check will always return the same result since `terminateResults ` does not change in the block.

Reply:
I'm actually wanting to refine the design more.  Ideally terminateResults should be changed to null outside of the thread to signal a graceful exit, but java scoping prevents that type of design.  Currently it exits from the interrupt signal which feels rough, but I haven't found a better option yet.  I considered using a signal value sent to the thread or logger in some was, but since it is performing logging operation this could introduce unexpected behavior for the user.